File: frontend/src/lib/services/mascotas.ts
import { apiFetch } from "../api";

export interface Pet {
    id: string;
    owner_id: string;
    name: string;
    species: string;
    breed: string | null;
    age_months: number | null;
    size: string | null;
    description: string | null;
    photo_url: string | null;
    is_active: boolean;
    adopted_from_listing_id: string | null;

    // Enrichment Fields
    is_vaccinated: boolean;
    is_sterilized: boolean;
    is_dewormed: boolean;
    temperament: string | null;
    energy_level: string | null;
    social_cats: boolean;
    social_dogs: boolean;
    social_children: boolean;
    weight_kg: number | null;
    microchip_number: string | null;
}

export async function getUserPets(userId: string): Promise<Pet[]> {
    return apiFetch<Pet[]>("mascotas", `/pets/user/${userId}`);
}

export async function getPetById(petId: string): Promise<Pet> {
    return apiFetch<Pet>("mascotas", `/pets/${petId}`);
}

export async function createPet(petData: Partial<Pet>): Promise<Pet> {
    return apiFetch<Pet>("mascotas", `/pets/`, {
        method: "POST",
        body: JSON.stringify(petData)
    });
}

export interface LostPet {
    id: string;
    user_id: string;
    pet_name: string;
    species: string;
    breed: string | null;
    description: string | null;
    last_seen_location: string;
    date_lost: string;
    contact_phone: string;
    image_url: string | null;
    is_found: boolean;
    created_at: string;
}

export interface PetfriendlyPlace {
    id: string;
    name: string;
    category: string;
    description: string | null;
    address: string;
    city: string | null;
    state: string | null;
    latitude: number | null;
    longitude: number | null;
    rating: number;
    image_url: string | null;
    created_by_user_id: string;
}

export async function getLostPets(isFound?: boolean): Promise<LostPet[]> {
    const params = isFound !== undefined ? `?is_found=${isFound}` : "";
    return apiFetch<LostPet[]>("mascotas", `/lost-pets/${params}`);
}

export async function createLostPet(
    pet: Omit<LostPet, "id" | "user_id" | "created_at">
): Promise<LostPet> {
    return apiFetch<LostPet>("mascotas", "/lost-pets/", {
        method: "POST",
        body: JSON.stringify(pet),
    });
}

export async function getPlaces(category?: string, city?: string): Promise<PetfriendlyPlace[]> {
    const params = new URLSearchParams();
    if (category) params.append("category", category);
    if (city) params.append("city", city);
    const qs = params.toString() ? `?${params.toString()}` : "";
    return apiFetch<PetfriendlyPlace[]>("mascotas", `/places/${qs}`);
}

export async function createPlace(
    place: Omit<PetfriendlyPlace, "id" | "created_by_user_id">
): Promise<PetfriendlyPlace> {
    return apiFetch<PetfriendlyPlace>("mascotas", "/places/", {
        method: "POST",
        body: JSON.stringify(place),
    });
}

File: backend/michicondrias_mascotas/app/models/mascotas.py
import uuid
from sqlalchemy import Column, String, Boolean, DateTime, Text, Float, Integer, JSON
from sqlalchemy.sql import func
from app.db.session import Base

class Pet(Base):
    __tablename__ = "pets"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    owner_id = Column(String, index=True, nullable=False) # User from Core
    name = Column(String, index=True, nullable=False)
    species = Column(String, nullable=False)
    breed = Column(String, nullable=True)
    age_months = Column(Integer, nullable=True)
    size = Column(String, nullable=True)
    description = Column(Text, nullable=True)
    photo_url = Column(String, nullable=True)
    is_active = Column(Boolean, default=True)
    
    # Enrichment Fields
    is_vaccinated = Column(Boolean, default=False)
    is_sterilized = Column(Boolean, default=False)
    is_dewormed = Column(Boolean, default=False)
    temperament = Column(Text, nullable=True)
    energy_level = Column(String, nullable=True) # Bajo, Medio, Alto
    social_cats = Column(Boolean, default=True)
    social_dogs = Column(Boolean, default=True)
    social_children = Column(Boolean, default=True)
    weight_kg = Column(Float, nullable=True)
    microchip_number = Column(String, nullable=True)
    gender = Column(String(20), nullable=True)
    gallery = Column(JSON, nullable=True)

    # Traceability: which adoption listing originated this pet record
    adopted_from_listing_id = Column(String, nullable=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

class LostPet(Base):
    __tablename__ = "lost_pets"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    user_id = Column(String, index=True, nullable=False) # Refers to users.id from core app.
    pet_name = Column(String, index=True, nullable=False)
    species = Column(String, nullable=False)
    breed = Column(String, nullable=True)
    description = Column(Text, nullable=True)
    last_seen_location = Column(String, nullable=False)
    date_lost = Column(DateTime(timezone=True), nullable=False)
    contact_phone = Column(String, nullable=False)
    image_url = Column(String, nullable=True)
    is_found = Column(Boolean, default=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

class PetfriendlyPlace(Base):
    __tablename__ = "petfriendly_places"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    name = Column(String, index=True, nullable=False)
    category = Column(String, index=True, nullable=False) # e.g. "restaurant", "park", "hotel"
    description = Column(Text, nullable=True)
    address = Column(String, nullable=False)
    city = Column(String, nullable=True)
    state = Column(String, nullable=True)
    latitude = Column(Float, nullable=True)
    longitude = Column(Float, nullable=True)
    rating = Column(Float, default=0.0)
    image_url = Column(String, nullable=True)
    created_by_user_id = Column(String, nullable=False)
